python/ray/serve/_private/deployment_state.py

```diff
@@ -2374,8 +2682,79 @@ def check_and_update_replicas(self):
                 self._replicas.add(ReplicaState.STOPPING, replica)
             else:
                 logger.info(f"{replica.replica_id} is stopped.")
+
+                # Release rank only after replica is successfully stopped
+                # This ensures rank is available during draining/graceful shutdown
+                replica_id = replica.replica_id.unique_id
+                self._rank_manager.release_rank(replica_id)
+                logger.info(
+                    f"Released rank from replica {replica_id} in deployment {self._id}"
+                )
+
                 self._autoscaling_state_manager.on_replica_stopped(replica.replica_id)
 
+        # After replica state updates, check rank consistency and perform minimal reassignment if needed
+        # This ensures ranks are continuous after lifecycle events
+        # Only do consistency check when deployment is stable (not during active updates)
+        # maybe this constraint need to be relaxed in the future. The implication is that
+        # if we delay the rank reassignment, the rank system will be in an invalid state
+        # for a longer period of time. Abrar made this decision because he is not confident
+        # about how rollouts work in the deployment state machine.
+        active_replicas = self._replicas.get()
+        if (
+            not self._has_outdated_version_replicas()
+            and len(active_replicas) == self._target_state.target_num_replicas
+        ):
+            if active_replicas:
+                replicas_to_reconfigure = (
+                    self._rank_manager.check_rank_consistency_and_reassign_minimally(
+                        active_replicas,
+                    )
+                )
+
+                # Reconfigure replicas that had their ranks reassigned
+                if replicas_to_reconfigure:
```

Comment:
not needed if we check for it already [here](https://github.com/ray-project/ray/pull/54938/files#diff-c4c2583a4c2f3a3c87ada6faebd0b2dfef404e165df133a11195be5d65fcb387R2734)?